### PR TITLE
[Access] Add initial register cache implementation

### PR DESCRIPTION
**Pull request by peterargue**
Closes: #5277

This PR adds a simple cache for register lookups on Access nodes. This improves the performance of scripts that lookup frequently queried registers, like common contract code.

### CODE REVIEW

File: storage/pebble/cache.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+const DefaultCacheSize = uint(1000)
+
+type CacheType int
```
Comment (peterargue) [inline storage/pebble/cache.go:16]: we can strip out the logic for selecting cache type after testing. initial benchmarking has shown that the `twoqueue` cache is the fastest with the lowest memory overhead

**Review by zhangchiqing — APPROVED** [review]

File: storage/pebble/registers_cache.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+// LatestHeight Gets the latest height of complete registers available
+func (c *RegistersCache) LatestHeight() uint64 {
+	return c.registers.LatestHeight()
+}
+
+// FirstHeight first indexed height found in the store, typically root block for the spork
+func (c *RegistersCache) FirstHeight() uint64 {
+	return c.registers.FirstHeight()
+}
```
Comment (janezpodhostnik) [inline storage/pebble/registers_cache.go:83]:
```suggestion
package pebble

import (
	"fmt"

	"github.com/pkg/errors"

	"github.com/onflow/flow-go/model/flow"
	"github.com/onflow/flow-go/module"
	"github.com/onflow/flow-go/storage"
)

const (
	registerResourceName = "registers"
)

type RegistersCache struct {
	*Registers
	cache     *ReadCache
}

var _ storage.RegisterIndex = (*RegistersCache)(nil)

// NewRegistersCache wraps a read cache around Get requests to a underlying Registers object.
func NewRegistersCache(registers *Registers, cacheType CacheType, size uint, metrics module.CacheMetrics) (*RegistersCache, error) {
	if size == 0 {
		return nil, errors.New("cache size cannot be 0")
	}

	cache, err := newReadCache(
		metrics,
		registerResourceName,
		cacheType,
		size,
		func(key string) (flow.RegisterValue, error) {
			return registers.lookupRegister([]byte(key))
		},
	)
	if err != nil {
		return nil, fmt.Errorf("could not create cache: %w", err)
	}

	return &RegistersCache{
		Registers: registers,
		cache:     cache,
	}, nil
}

// Get returns the most recent updated payload for the given RegisterID.
// "most recent" means the updates happens most recent up the given height.
//
// For example, if there are 2 values stored for register A at height 6 and 11, then
// GetPayload(13, A) would return the value at height 11.
//
// - storage.ErrNotFound if no register values are found
// - storage.ErrHeightNotIndexed if the requested height is out of the range of stored heights
func (c *RegistersCache) Get(
	reg flow.RegisterID,
	height uint64,
) (flow.RegisterValue, error) {
	return c.cache.Get(newLookupKey(height, reg).String())
}
```

File: storage/pebble/registers_cache.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+var _ storage.RegisterIndex = (*RegistersCache)(nil)
+
+// NewRegistersCache wraps a read cache around Get requests to a underlying Registers object.
+func NewRegistersCache(registers *Registers, cacheType CacheType, size uint, metrics module.CacheMetrics) (*RegistersCache, error) {
+	if size == 0 {
+		return nil, errors.New("cache size cannot be 0")
+	}
+
+	cache, err := newReadCache(
```
Comment (janezpodhostnik) [inline storage/pebble/registers_cache.go:30]: The responsibility of creating a cache doesnt need to be in here... 
The new function could just be: `NewRegistersCache(registers *Registers, cache     *ReadCache)`

File: storage/pebble/cache.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	resource, cached := c.cache.Get(key)
+	if cached {
+		c.metrics.CacheHit(c.resource)
+		return resource, nil
+	}
+
+	// get it from the database
+	resource, err := c.retrieve(key)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
```
Comment (janezpodhostnik) [inline storage/pebble/cache.go:147]: `storage.ErrNotFound` could potentially be cached, since it wont change over time.

**Review by janezpodhostnik — CHANGES_REQUESTED** [review]
I think a useful feature to have (for the future) is an admin command to clear this cache. Might bee good if we are ever debugging any issues.

File: storage/pebble/registers_cache.go (hunk 3 above)
Comment (janezpodhostnik) [reply storage/pebble/registers_cache.go:30]: I don't fell strongly about this.

Comment (peterargue) [reply storage/pebble/registers_cache.go:30]: ok I'll leave it for now. I think we can evolve this some more to make the API more general purpose as we add more types to pebble.

**Pull request merged**